File: src/mobile/crypto.ts
/**
 * Phone-side cryptography — the WebCrypto mirror of src/main/remote/crypto.ts.
 *
 * Same constants from @shared/remote-protocol, same wire format, so the two
 * sides interoperate by construction: the pairing secret from the QR derives
 * the HKDF frame key; every application frame is AES-256-GCM sealed. The
 * secret itself is read from the URL fragment exactly once and cleared, never
 * stored, never sent.
 */

import {
  FRAME_KEY_INFO,
  FRAME_KEY_SALT,
  PAIRING_PROOF_CONTEXT,
  type SealedFrame,
} from '@shared/remote-protocol'

const encoder = new TextEncoder()
const decoder = new TextDecoder()

/** Bytes with a plain ArrayBuffer backing — what WebCrypto accepts everywhere. */
type Bytes = Uint8Array<ArrayBuffer>

const bytes = (length: number): Bytes => new Uint8Array(new ArrayBuffer(length))

const toHex = (data: Uint8Array): string =>
  [...data].map((b) => b.toString(16).padStart(2, '0')).join('')

const toBase64 = (data: Uint8Array): string => {
  let binary = ''
  for (const byte of data) binary += String.fromCharCode(byte)
  return btoa(binary)
}

const fromBase64 = (text: string): Bytes => {
  const binary = atob(text)
  const out = bytes(binary.length)
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i)
  return out
}

/** HMAC-SHA256(secret, PAIRING_PROOF_CONTEXT) as hex — the pairing proof. */
export async function pairingProof(secret: string): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw',
    encoder.encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  )
  const mac = await crypto.subtle.sign('HMAC', key, encoder.encode(PAIRING_PROOF_CONTEXT))
  return toHex(new Uint8Array(mac))
}

/** HKDF-SHA256(secret) → 256-bit AES-GCM frame key (raw bytes, base64-storable). */
export async function deriveFrameKey(secret: string): Promise<Bytes> {
  const material = await crypto.subtle.importKey('raw', encoder.encode(secret), 'HKDF', false, [
    'deriveBits',
  ])
  const bits = await crypto.subtle.deriveBits(
    {
      name: 'HKDF',
      hash: 'SHA-256',
      salt: encoder.encode(FRAME_KEY_SALT),
      info: encoder.encode(FRAME_KEY_INFO),
    },
    material,
    256
  )
  return new Uint8Array(bits)
}

const importAesKey = (raw: Bytes): Promise<CryptoKey> =>
  crypto.subtle.importKey('raw', raw, 'AES-GCM', false, ['encrypt', 'decrypt'])

/** Raw-key ⇄ base64 (identity storage in localStorage). */
export const keyToBase64 = (key: Bytes): string => toBase64(key)
export const keyFromBase64 = (text: string): Bytes => fromBase64(text)

/** Seals one JSON-serializable frame: AES-256-GCM with a fresh 12-byte nonce. */
export async function sealFrame(key: Bytes, payload: unknown): Promise<SealedFrame> {
  const nonce = crypto.getRandomValues(bytes(12))
  const cipher = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: nonce },
    await importAesKey(key),
    encoder.encode(JSON.stringify(payload))
  )
  return { t: 'sec', n: toBase64(nonce), ct: toBase64(new Uint8Array(cipher)) }
}

/** Opens one sealed frame; throws on tampering or a wrong key. */
export async function openFrame<T>(key: Bytes, frame: SealedFrame): Promise<T> {
  const plain = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: fromBase64(frame.n) },
    await importAesKey(key),
    fromBase64(frame.ct)
  )
  return JSON.parse(decoder.decode(plain)) as T
}

File: mobile_flutter/lib/protocol/crypto.dart
/// Phone-side cryptography — the Dart mirror of src/remote/crypto.ts (which is
/// itself the WebCrypto mirror of src/main/remote/crypto.ts).
///
/// Same constants, same wire format, so all three sides interoperate by
/// construction: the pairing secret from the QR derives the HKDF frame key;
/// every application frame is AES-256-GCM sealed. The sealed frame's `ct` is
/// WebCrypto's ciphertext||tag layout (tag last, 16 bytes) — that layout is
/// part of the wire contract and must not change.
library;

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'constants.dart';

final AesGcm _aesGcm = AesGcm.with256bits();

/// Cryptographically secure random bytes (frame nonces, request ids).
Uint8List randomBytes(int length) {
  final rng = Random.secure();
  return Uint8List.fromList(List<int>.generate(length, (_) => rng.nextInt(256)));
}

String bytesToHex(List<int> bytes) =>
    bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();

/// HMAC-SHA256(secret, pairingProofContext) as hex — the pairing proof.
/// The secret itself never crosses the wire; this proof does.
Future<String> pairingProof(String secret) async {
  final mac = await Hmac.sha256().calculateMac(
    utf8.encode(pairingProofContext),
    secretKey: SecretKey(utf8.encode(secret)),
  );
  return bytesToHex(mac.bytes);
}

/// HKDF-SHA256(secret) → 256-bit AES-GCM frame key (raw bytes, storable
/// base64). Salt/info are the fixed public constants — all entropy is in the
/// 256-bit pairing secret.
Future<Uint8List> deriveFrameKey(String secret) async {
  final hkdf = Hkdf(hmac: Hmac.sha256(), outputLength: 32);
  final key = await hkdf.deriveKey(
    secretKey: SecretKey(utf8.encode(secret)),
    nonce: utf8.encode(frameKeySalt),
    info: utf8.encode(frameKeyInfo),
  );
  return Uint8List.fromList(await key.extractBytes());
}

/// Raw key bytes ⇄ base64 (identity storage).
String keyToBase64(List<int> key) => base64Encode(key);
Uint8List keyFromBase64(String text) => base64Decode(text);

/// Seals one JSON-serializable frame: AES-256-GCM with a fresh 12-byte nonce.
/// Returns the relay-visible `{t:'sec', n, ct}` map, `ct` = ciphertext||tag.
Future<Map<String, dynamic>> sealFrame(SecretKey key, Object? payload) async {
  final nonce = randomBytes(12);
  final box = await _aesGcm.encrypt(
    utf8.encode(jsonEncode(payload)),
    secretKey: key,
    nonce: nonce,
  );
  final sealed = BytesBuilder()..add(box.cipherText)..add(box.mac.bytes);
  return {'t': 'sec', 'n': base64Encode(nonce), 'ct': base64Encode(sealed.toBytes())};
}

/// Opens one sealed frame; throws on tampering or a wrong key.
Future<Object?> openFrame(SecretKey key, Map<String, dynamic> frame) async {
  final nonce = base64Decode(frame['n'] as String);
  final ct = base64Decode(frame['ct'] as String);
  if (ct.length < 16) {
    throw const FormatException('sealed frame too short');
  }
  final box = SecretBox(
    ct.sublist(0, ct.length - 16),
    nonce: nonce,
    mac: Mac(ct.sublist(ct.length - 16)),
  );
  final plain = await _aesGcm.decrypt(box, secretKey: key);
  return jsonDecode(utf8.decode(plain));
}

/// A random RFC 4122 v4 UUID, as the request id (matches the web client's
/// crypto.randomUUID — the desktop treats it as an opaque correlation id).
String randomUuid() {
  final b = randomBytes(16);
  b[6] = (b[6] & 0x0f) | 0x40;
  b[8] = (b[8] & 0x3f) | 0x80;
  final hex = bytesToHex(b);
  return '${hex.substring(0, 8)}-${hex.substring(8, 12)}-'
      '${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20)}';
}
